### ISSUE TITLE
Bug: [restrict-template-expressions] should check if superclass is ignored

### ISSUE THREAD

**Issue by joshkel** [typescript-eslint/typescript-eslint#11759]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=MYGwhgzhAEBCkFNoG9oF8BQpIwCIICcBLANwQBNoEAPAFwQDtyZ4IlVNgB7BiW6AGZcu0ALzQGCAO5xEACgCUAbiw8%2B0AEZgCYidOj5iZcopUADACTIhXNFa0E0ZpUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aFfKsu-RAW2LwAhn3oAPYh2TJKAeyaoMAbXDYciaNDnRIAGjXrgUYfHhyA7ujDLIAIWHJEkALoBfQ2BdqPboA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
class Base { }
class Derived extends Base { }
const foo = new Base();
const bar = new Derived();
`${foo}${bar}`;
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": [
      "error",
      { "allow": ["Base"]}
    ]
  }
}
```

### tsconfig

```jsonc
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": [
      "error",
      { "allow": ["Base"]}
    ]
  }
}
```

### Expected Result

No warnings - if `Base` is allowed and every instance of `Derived` is also a `Base`, then `Derived` should be allowed. (This would also be consistent with the `no-base-to-string` rule - see #11520.)

### Actual Result

> Invalid type "Derived" of template literal expression. 5:10 - 5:13

**Comment by Higangssh** [comment]
Hi there! I've submitted PR #11764 to fix this issue.

  I've implemented recursive base type checking to allow derived classes when their base class is in the allow list.
   I followed the same pattern as `no-base-to-string` for consistency.

  If this approach looks reasonable, I'd be grateful for a review. If there's a better way to handle this, I'm happy
   to revise. Thank you!

**Comment by JoshuaKGoldberg** [comment]
Agreed with the parallel to #11520. Accepting PRs!

@Higangssh I appreciate the enthusiasm! A couple of tips:

* I'd suggest waiting until an issue has been noted by a maintainer as accepting PRs before sending yours. This one was straightforward and accepted in this case, but you never know when something ends up being surprisingly nuanced and not accepted. I'd hate to see you spend time on a change that can't be accepted.
* There's no need to comment in this issue; properly linking the PR in its description means we'll see it, see the explanation, and give it a review when we can.

**Comment by Higangssh** [comment]
@JoshuaKGoldberg Thank you for taking the time to guide me! 
I really appreciate the advice and will follow that process for future contributions. Thanks again for reviewing!
